lib/auth/saml.go

```diff
@@ -148,21 +178,53 @@ func (a *Server) getSAMLProvider(conn types.SAMLConnector) (*saml2.SAMLServicePr
 	return serviceProvider, nil
 }
 
-func (a *Server) calculateSAMLUser(connector types.SAMLConnector, assertionInfo saml2.AssertionInfo, request *services.SAMLAuthRequest) (*createUserParams, error) {
-	var err error
+type ssoDiagContext struct {
+	authKind                string
+	createSSODiagnosticInfo func(ctx context.Context, authKind string, authRequestID string, info types.SSODiagnosticInfo) error
```

Comment:
This field might warrant a godoc explaining how it's used.

And come to think of it, godocs on other fields might be useful too.